target/linux/rtl838x/files-5.4/drivers/net/dsa/rtl838x_phy.c

```diff
@@ -0,0 +1,1287 @@
+#define FIRMWARE_838X_8218b_1	"rtl838x_phy/rtl838x_8218b.fw"
+
+static const struct firmware rtl838x_8380_fw;
+static const struct firmware rtl838x_8214fc_fw;
+static const struct firmware rtl838x_8218b_fw;
+
+struct rtl838x_phy_priv {
+	char *name;
+};
+
+static int read_phy(u32 port, u32 page, u32 reg, u32 *val)
+{
+	if (soc_info.family == RTL8390_FAMILY_ID)
+		return rtl839x_read_phy(port, page, reg, val);
+	else
+		return rtl838x_read_phy(port, page, reg, val);
+}
+
+static int write_phy(u32 port, u32 page, u32 reg, u32 val)
+{
+	if (soc_info.family == RTL8390_FAMILY_ID)
+		return rtl839x_write_phy(port, page, reg, val);
+	else
+		return rtl838x_write_phy(port, page, reg, val);
+}
+
+static void rtl8380_int_phy_on_off(int mac, bool on)
+{
+	u32 val;
+
+	read_phy(mac, 0, 0, &val);
+	if (on)
+		write_phy(mac, 0, 0, val & ~(1 << 11));
+	else
+		write_phy(mac, 0, 0, val | (1 << 11));
+}
+
+static void rtl8380_rtl8214fc_on_off(int mac, bool on)
+{
+	u32 val;
+
+	/* fiber ports */
+	write_phy(mac, 4095, 30, 3);
+	read_phy(mac, 0, 16, &val);
+	if (on)
+		write_phy(mac, 0, 16, val & ~(1 << 11));
+	else
+		write_phy(mac, 0, 16, val | (1 << 11));
+
+	/* copper ports */
+	write_phy(mac, 4095, 30, 1);
+	read_phy(mac, 0, 16, &val);
+	if (on)
+		write_phy(mac, 0xa40, 16, val & ~(1 << 11));
+	else
+		write_phy(mac, 0xa40, 16, val | (1 << 11));
+}
+
+static void rtl8380_phy_reset(int mac)
+{
+	u32 val;
+
+	read_phy(mac, 0, 0, &val);
+	write_phy(mac, 0, 0, val | (0x1 << 15));
+}
+
+void rtl8380_sds_rst(int mac)
+{
+	u32 offset = (mac == 24) ? 0 : 0x100;
+
+	sw_w32_mask(1 << 11, 0, RTL8380_SDS4_FIB_REG0 + offset);
+	sw_w32_mask(0x3, 0, RTL838X_SDS4_REG28 + offset);
+	sw_w32_mask(0x3, 0x3, RTL838X_SDS4_REG28 + offset);
+	sw_w32_mask(0, 0x1 << 6, RTL838X_SDS4_DUMMY0 + offset);
+	sw_w32_mask(0x1 << 6, 0, RTL838X_SDS4_DUMMY0 + offset);
+	pr_info("SERDES reset: %d\n", mac);
+}
+
+static struct fw_header *
+rtl838x_request_fw(struct phy_device *phydev, const struct firmware *fw,
+		   const char *name)
+{
+	struct device *dev = &phydev->mdio.dev;
+	int err;
+	struct fw_header *h;
+	uint32_t checksum, my_checksum;
+
+	err = request_firmware(&fw, name, dev);
+	if (err < 0)
+		goto out;
+
+	if (fw->size < sizeof(struct fw_header)) {
+		pr_err("Firmware size too small.\n");
+		err = -EINVAL;
+		goto out;
+	}
+
+	h = (struct fw_header *) fw->data;
+	pr_info("Firmware loaded. Size %d, magic: %08x\n", fw->size, h->magic);
+
+	if (h->phy != 0x83800000) {
+		pr_err("Wrong firmware file: PHY mismatch.\n");
+		goto out;
+	}
+
+	checksum = h->checksum;
+	h->checksum = 0;
+	my_checksum = ~crc32(0xFFFFFFFFU, fw->data, fw->size);
+	if (checksum != my_checksum) {
+		pr_err("Firmware checksum mismatch.\n");
+		err = -EINVAL;
+		goto out;
+	}
+	h->checksum = checksum;
+
+	return h;
+out:
+	dev_err(dev, "Unable to load firmware %s (%d)\n", name, err);
+	return NULL;
+}
+
+static int rtl8390_configure_generic(struct phy_device *phydev)
+{
+	u32 val, phy_id;
+//	int i, p, ipd_flag;
```

Comment:
ERROR: do not use C99 // comments

Can be removed?

Reply:
Will be removed.